system/lib/wasmfs/syscalls.cpp

```diff
@@ -889,4 +889,53 @@ long __syscall_rename(long old_path, long new_path) {
   return 0;
 }
 
+long __syscall_symlink(char* old_path, char* new_path) {
+  auto pathParts = splitPath(new_path);
+
+  long err;
+  auto parsedPath = getParsedPath(pathParts, err);
+
+  if (!parsedPath.parent) {
+    return err;
+  }
+
+  if (pathParts.back().size() > WASMFS_NAME_MAX) {
+    return -ENAMETOOLONG;
+  }
+
+  if (parsedPath.child) {
+    return -EEXIST;
+  }
+
+  auto backend = parsedPath.parent->unlocked()->getBackend();
+  auto created = backend->createSymlink(old_path);
+  parsedPath.parent->setEntry(pathParts.back(), created);
+
+  return 0;
+}
+
+long __syscall_readlink(char* path, char* buf, size_t bufSize) {
+  auto pathParts = splitPath(path);
+
+  long err;
+  auto parsedPath = getParsedPath(pathParts, err);
+  if (!parsedPath.parent) {
+    return err;
+  }
+  if (!parsedPath.child) {
+    return -ENOENT;
+  }
+  if (!parsedPath.child->is<Symlink>()) {
+    return -EINVAL;
+  }
+
+  auto symlink = parsedPath.child->dynCast<Symlink>();
+  auto& target = symlink->locked().getTarget();
```

Comment:
Holding a reference that outlives the lock needed to acquire it seems fishy. But the targets of symlinks are immutable, so we can make this less fishy by returning a `const std::string&` from `getTarget()`. We could go even further by putting `getTarget()` on `Symlink` directly rather than on `Symlink::Handle` since the returned data is immutable and therefore thread-safe.

Reply:
Ah, very good catch, and very good idea! Modified to do it that way now.